### DISCUSSION TITLE
Feedback: NodeJS + Typescript

### DISCUSSION THREAD

**Discussion by alexcastillo** [firebase/firebase-js-sdk#4468] [General]
Hello Firebase team!

Thank you so much for the amazing effort to get this massive refactor out. 

I've gone through the documentation and love the direction you've taken to solve the modularization problem. Well done!

I used the `firebase@exp` branch for a new project with the following setup, and I have some feedback.

## Setup

 - node.js v15.4.0
 - typescript v4.1.5
 - firebase 0.900.12

tsconfig.json

```json
{
  "compilerOptions": {
    "target": "es2019",
    "module": "commonjs",
    "outDir": "dist",
    "lib": ["es2020", "dom"],
    "strict": true,
    "declaration": true,
    "esModuleInterop": true,
    "noImplicitAny": false,
    "downlevelIteration": true
  },
  "exclude": ["node_modules", "dist", "scripts", "examples"]
}

```

### 1. Unable to build with Typescript

I'm unable to build with Typescript (`tsc`) due to the following error:

![image](https://user-images.githubusercontent.com/1181666/107817929-08e6c300-6d34-11eb-94f2-5b9e53ea38d3.png)

```
node_modules/@firebase/storage-types/exp/index.d.ts:18:23 - error TS2305: Module '"../../app-types"' has no exported member '_FirebaseService'.

18 import { FirebaseApp, _FirebaseService } from '@firebase/app-types';
```

### 2. Type error when using initializeApp

```
Argument of type is not assignable to parameter of type 'import("/node_modules/@firebase/app-types/index").FirebaseApp'.
```

> Workaround is to use `any`: export const firebaseApp: any = initializeApp(config);

### 3. Unable to use firestore without firestore/lite

When doing a compound query with Firestore, I get the following runtime error:

```
@firebase/firestore: Firestore (8.2.5): INTERNAL UNHANDLED ERROR:  Error: ENOENT: no such file or directory, open 'node_modules/@firebase/firestore/dist/src/protos/google/firestore/v1/firestore.proto'
    at Object.openSync (node:fs:495:3)
    at Object.readFileSync (node:fs:396:35)
    at fetch (node_modules/protobufjs/src/root.js:172:34)
    at Root.load (node_modules/protobufjs/src/root.js:206:13)
    at Root.loadSync (node_modules/protobufjs/src/root.js:247:17)
    at Object.loadSync (node_modules/@grpc/proto-loader/build/src/index.js:223:29)
    at loadProtos (node_modules/@firebase/firestore/src/platform/node/load_protos.ts:46:29)
    at newConnection (node_modules/@firebase/firestore/src/platform/node/connection.ts:28:18)
    at OnlineComponentProvider.createDatastore (node_modules/@firebase/firestore/src/core/component_provider.ts:367:24)
    at OnlineComponentProvider.<anonymous> (node_modules/@firebase/firestore/src/core/component_provider.ts:335:27)
Error: ENOENT: no such file or directory, open 'node_modules/@firebase/firestore/dist/src/protos/google/firestore/v1/firestore.proto'
    at Object.openSync (node:fs:495:3)
    at Object.readFileSync (node:fs:396:35)
    at fetch (node_modules/protobufjs/src/root.js:172:34)
    at Root.load (node_modules/protobufjs/src/root.js:206:13)
    at Root.loadSync (node_modules/protobufjs/src/root.js:247:17)
    at Object.loadSync (node_modules/@grpc/proto-loader/build/src/index.js:223:29)
    at loadProtos (node_modules/@firebase/firestore/src/platform/node/load_protos.ts:46:29)
    at newConnection (node_modules/@firebase/firestore/src/platform/node/connection.ts:28:18)
    at OnlineComponentProvider.createDatastore (node_modules/@firebase/firestore/src/core/component_provider.ts:367:24)
    at OnlineComponentProvider.<anonymous> (node_modules/@firebase/firestore/src/core/component_provider.ts:335:27)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! start: `ts-node ./src/index.ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

I also get a good amount of typescript errors when not using /lite.

![image](https://user-images.githubusercontent.com/1181666/107818357-b8239a00-6d34-11eb-8256-6a702a4b2637.png)

> Workaround is to use /lite when importing:

```
import { getFirestore, collection, query, where, getDocs } from "firebase/firestore/lite";
```

### 4. Unable to get user session via stsTokenManager

In legacy firebase sdk, the only way to persist the auth session in node is with the following workaround:

```js
const userData = JSON.parse(userJson)
const user = new firebase.User(userData, userData.stsTokenManager, userData)
firebase.auth().updateCurrentUser(user)
```
Source: https://github.com/firebase/firebase-js-sdk/issues/1874#issuecomment-549085119

In firebase exp, there doesn't seem to be a way to get `User` from either `firebase/app` or `firebase/auth`.

> No workaround

I hope this is helpful, keep up the amazing work!

Alex

**Comment by Feiyang1** [comment]
Thanks for the feedback!

> 1. Unable to build with Typescript
> 2. Type error when using initializeApp

For the first 2 issues, it seems you have a production version of `@firebase/app-types` installed. Have you removed the production firebase packages from your dependencies? If you did, a clean reinstall (delete node_modules then npm i) should solve the issue.

> 3. Unable to use firestore without firestore/lite

Good catch on the runtime error. We are looking at the wrong location for the protos. Opened https://github.com/firebase/firebase-js-sdk/issues/4480 to track

We are already aware of the typings [issue](https://github.com/firebase/firebase-js-sdk/issues/4441), and made a fix (not released yet). Once a new version is released, the issue will go away.

> 4. Unable to get user session via stsTokenManager

@samhorlbeck Any thoughts?

**Reply by sam-gc** [reply]
This is interesting. We implemented `User` to be an interface with an underlying class that's not exposed. I'm not super keen changing the way this works (since it has ramifications for the compatibility layer), but perhaps we should add a function for constructing a user?

**Reply by Feiyang1** [reply]
I think we should since it's a supported use case in the current SDK. Adding a function sounds good to me if exporting the class may introduce interoperability issues between the new SDK and the compat layer. Can you explain what issues will there be if we export the class?

**Reply by alexcastillo** [reply]
Hi @sam-gc, @Feiyang1,

Any progress on item no. 4? This issue prevents node and electron apps from persisting auth.

**Reply by alexcastillo** [reply]
Looks like there's a PR open (#4484) that addresses this 🎉
